### ISSUE TITLE
Accessibility - link to label(s) of existing <select> element

### ISSUE THREAD

**Issue by binary-koan** [select2/select2#3835]
Also mentioned in #3144 

To be properly accessible to screen readers, form elements should be linked properly to their labels - this becomes a problem when you have something like:

``` html
<select class="select2" id="my-select">...</select>
<label for="my-select">...</label>
```

If you turn the select box into a select2 it will no longer be linked to its label because the original `<select>` is hidden.

The solution we've implemented in Mahara is: when a select2 is constructed, search for labels which link to it. Give them IDs if they don't already have one, and set the `aria-labelledby` attribute of the select2 to be a list of those IDs.

I'm not sure whether that's the best thing to do since it modifies other elements on the page (adding IDs to the labels) but it would be nice if something similar could be implemented, so that select2 really is a seamless drop-in select box replacement.

**Comment by ivanmcgregor** [comment]
As labels link to the next id in the DOM that has the referred to ID, could a simple solution work, where the `<span>` with the select2 is inserted before the `<select>`? 

I am too little into select2 or screen readers but maybe my two cents here can be of help.

**Comment by bkosborne** [comment]
The proposed solution of using the "labelledby" attribute works, at least for single selects. I put this near the top of `SingleSelection.prototype.bind` function:

```
    // Associate the selection element with the original label element for
    // better accessibility. To do this, we need to get the ID of the
    // label and set it as the 'aria-labelledby' attribute on the selection.
    var elementId = this.$element.attr('id');
    if (elementId) {
      var $label = $(document).find('label[for="'+elementId+'"]');
      if ($label.length > 0) {
        var labelId = $label.attr('id');
        // If the label doesn't have an ID, generate one based on the element ID.
        if (!labelId) {
          labelId = 'select2-label-' + elementId;
          $label.attr('id', labelId);
        }
        this.$selection.attr('aria-labelledby', labelId);
      }
    }
```

Still looking into multi select. I think the labelledby would have to be added to the search input field.
